fix(stepfunctions-tasks): cannot specify part of execution data or task context as input to the `RunLambda` service integration (#7428)
aws/aws-cdk: packages/@aws-cdk/aws-stepfunctions/README.md

#### Lambda example

[Invoke](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) a Lambda function.

You can specify the input to your Lambda function through the `payload` attribute.
By default, no payload is specified so Step Functions invokes Lambda with the empty
object `{ }` as input.

> The following snippet invokes a Lambda Function with the task context as the input

Author: ```suggestion
The following snippet invokes a Lambda Function with the state input as the payload
```

by referencing the `$` path.

```ts
new sfn.Task(this, 'Invoke with task context', {
  task: new tasks.RunLambdaTask(myLambda, {
    payload: sfn.TaskInput.fromDataAt('$'),
  }),
});
```

> When a function is invoked, the Lambda service sends back the following
> [response elements](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_ResponseElements)

Reviewer: ```suggestion
When a function is invoked, the Lambda service sends  [these response
elements](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_ResponseElements)
back.
```

⚠️ The response from the Lambda function is in an attribute called `Payload`

The following snippet invokes a Lambda Function by referencing the `$.Payload` path
to reference the output of a Lambda executed before it.

```ts
new sfn.Task(this, 'Invoke with empty payload', {
  task: new tasks.RunLambdaTask(myLambda),
});

new sfn.Task(this, 'Invoke with payload field in the state input', {
  task: new tasks.RunLambdaTask(myOtherLambda, {
    payload: sfn.TaskInput.fromDataAt('$.Payload'),
  }),
});
```

> The following snippet invokes a Lambda and sets the task output to only include

Reviewer: "only include" - as opposed to what else?

Author: as opposed to all of the response elements `StatusCode`, `ExecutionResult`, etc... https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_ResponseElements

so this only returns the Lambda return which will be in the `Payload` attribute. I think this is covered in the blurb above though?

the Lambda function response.

```ts
new sfn.Task(this, 'Invoke and set function response as task output', {
  task: new tasks.RunLambdaTask(myLambda, {
    payload: sfn.TaskInput.fromDataAt('$'),
  }),
  outputPath: '$.Payload',
});
```

You can have Step Functions pause a task, and wait for an external process to
return a task token. Read more about the [callback pattern](https://docs.aws.amazon.com/step-functions/latest/dg/callback-task-sample-sqs.html#call-back-lambda-example)

To use the callback pattern, set the `token` property on the task. Call the Step
Functions `SendTaskSuccess` or `SendTaskFailure` APIs with the token to
indicate that the task has completed and the state machine should resume execution.

The following snippet invokes a Lambda with the task token as part of the input
to the Lambda.

```ts
  const task = new sfn.Task(stack, 'Invoke with callback', {
    task: new tasks.RunLambdaTask(myLambda, {
      integrationPattern: sfn.ServiceIntegrationPattern.WAIT_FOR_TASK_TOKEN,
      payload: {
        token: sfn.Context.taskToken,
        input: sfn.TaskInput.fromDataAt('$.someField'),
      }
    })
  });
```

> ⚠️ The task will pause until it receives that task token back with a SendTaskSuccess or SendTaskFailure call. Read [more](https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html#connect-wait-token)

Reviewer: ```suggestion
⚠️ The task will pause until it receives that task token back with a `SendTaskSuccess` or `SendTaskFailure`
call. Learn more about [Callback with the Task
Token](https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html#connect-wait-token).
```

#### Lambda example

[Invoke](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) a Lambda function.

You can specify the input to your Lambda function through the `payload` attribute.
By default, no payload is specified so Step Functions invokes Lambda with the empty
object `{ }` as input.

The following snippet invokes a Lambda Function with the task context as the input
by referencing the `$` path.

```ts
new sfn.Task(this, 'Invoke with task context', {
  task: new tasks.RunLambdaTask(myLambda, {
    payload: sfn.TaskInput.fromDataAt('$'),
  }),
});
```

When a function is invoked, the Lambda service sends back the following
[response elements](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_ResponseElements)

⚠️ The response from the Lambda function is in an attribute called `Payload`

The following snippet invokes a Lambda Function by referencing the `$.Payload` path
to reference the output of a Lambda executed before it.

```ts

> new sfn.Task(this, 'Invoke with task context', {

Author: ```suggestion
new sfn.Task(this, 'Invoke with empty payload', {
```

  task: new tasks.RunLambdaTask(myLambda),
});

> new sfn.Task(this, 'Invoke with output from another Lambda', {

Reviewer 2: How about `Invoke with output from the previous task` or `Invoke with the Payload field in the state input`?

It's kind of presented here as a generic way to invoke Lambda but only applies if you've structured your states exactly in this way. I can see users who don't understand input/output processing to just expect it to work all the time. The second option makes sense on its own.

> new sfn.Task(this, 'Invoke with output from another Lambda', {

Author: ```suggestion
new sfn.Task(this, 'Invoke with payload field in the state input', {
```

  task: new tasks.RunLambdaTask(myOtherLambda, {
    payload: sfn.TaskInput.fromDataAt('$.Payload'),
  }),
});
```

The following snippet invokes a Lambda and sets the task output to only include
the Lambda function response.

```ts
new sfn.Task(this, 'Invoke and set function response as task output', {

>   task: new tasks.RunLambdaTask(checkJobStateLambda, {

Author: ```suggestion
  task: new tasks.RunLambdaTask(myLambda, {
```

    payload: sfn.TaskInput.fromDataAt('$'),
  }),
  outputPath: '$.Payload',
});
```

You can have Step Functions pause a task, and wait for an external process to
return a task token. Read more about the [callback pattern](https://docs.aws.amazon.com/step-functions/latest/dg/callback-task-sample-sqs.html#call-back-lambda-example)

To use the callback pattern, set the `token` property on the task and have the

> Lambda function Lambda function call the Step Functions API `SendTaskSuccess`

Reviewer 2: typo: Lambda function x2
typo: Step Functions `SendTaskSuccess` or `SendTaskFailure` APIs

This should just say that *something* needs to callback. It's more typical for a Lambda invoked with `.waitForTaskToken` to send the token elsewhere for processing.

> To use the callback pattern, set the `token` property on the task and have the
> Lambda function Lambda function call the Step Functions API `SendTaskSuccess`
> or `SendTaskFailure` API with the token to indicate that the task has completed
> and the state machine should resume execution.

Author: ```suggestion
To use the callback pattern, set the `token` property on the task. Call the Step
Functions `SendTaskSuccess` or `SendTaskFailure` APIs with the token to
indicate that the task has completed and the state machine should resume execution.
```

The following snippet invokes a Lambda with the task token as part of the input
to the Lambda.

```ts
  const task = new sfn.Task(stack, 'Invoke with callback', {
    task: new tasks.RunLambdaTask(myLambda, {
      integrationPattern: sfn.ServiceIntegrationPattern.WAIT_FOR_TASK_TOKEN,
      payload: {
        token: sfn.Context.taskToken,
        input: sfn.TaskInput.fromDataAt('$.someField'),
      }
    })
  });
```

> ⚠️ The Lambda function should call `SendTaskSuccess` or `SendTaskFailure` with the

Reviewer 2: `.waitForToken` would typically be used with Lambda to send the token somewhere else (an on-prem service, S3, etc) for processing. Calling `SendTaskSuccess/Failure` in the same function that received the token probably isn't the typical use case. If the work can be started and completed within a single function invocation, it makes more sense to just use the Request Response integration pattern.

The example you linked in this section is for a sample project where the token is actually sent to SQS using `arn:aws:states:::sqs:sendMessage.waitForTaskToken`. In that example, the Lambda function is triggered by SQS outside of the state machine. I think the conceptual documentation Wait for Task Token is more appropriate here. 

https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html#connect-wait-token

> ⚠️ The Lambda function should call `SendTaskSuccess` or `SendTaskFailure` with the
> token provided as input or the State Machine will not resume.

Author: ```suggestion
⚠️ The task will pause until it receives that task token back with a SendTaskSuccess or SendTaskFailure call. Read [more](https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html#connect-wait-token)
```
